src/StoreApi/Utilities/CartController.php

```diff
@@ -207,6 +207,24 @@ public function set_cart_item_quantity( $item_id, $quantity = 1 ) {
 		$cart->set_quantity( $item_id, $quantity );
 	}
 
+	/**
+	 * Sets the custom cart item data for a given cart item.
+	 *
+	 * @param string $item_id Cart item id.
+	 * @param array  $data {
+	 *    Custom data to set for the cart item.
+	 *
+	 *    @type string $namespace Namespace for the custom data.
+	 *    @type string $key       Key for the custom data.
+	 *    @type mixed  $value     Value for the custom data.
+	 *
+	 * @return void
+	 */
+	public function set_cart_item_custom_data( $item_id, $data ) {
+		$session_key = $item_id . '_' . $data['namespace'] . '_' . $data['key'];
+		wc()->session->set( $session_key, $data['value'] );
```

Comment:
Is this safe to be session only? I'm just wondering how tied it needs to be to cart data, since those get stored into the DB. If its temporary data maybe it doesn't matter so much.

Reply:
We'd need to save it to the order when it's placed, do you know off the top of your head if it's possible to add metadata to order items when saving the order?

Reply:
Might involve pulling some logic out of core into the store API (update_line_items_from_cart method).